### ISSUE TITLE
Interest in a fork of the Lua extension for VS Code?

### ISSUE THREAD

**Issue by hahn-kev** [FAForever/fa#3858]
Hi, I've been watching some of the coding videos on youtube and I noticed that the Lua editing experience isn't that great. I also saw in the deployment setup that using the Lua extension is not recommended. Text here

> Note that we do not recommend the Lua language server. The game uses a slightly adjusted version of Lua. The syntax doesn't match. We also have a different import system. The Lua language server generates dozens of errors. And it is not equipped to work with our import system.

Would there be interest if I were to take an existing extension (with permission) and fork it for use for working on FA?

For reference I've been working on an extension for modding another game, you can find the repo [here](https://github.com/hahn-kev/ftl-xml-extension) if you would like. I'd love to help support the devs by working on an extension for them to help with errors and code quality etc.

**Comment by Garanas** [comment]
Absolutely! Once the extension is taking shape we can actively promote it using the same coding videos.

@Askaholic has already done some work in the past to adjust a Lua distribution to match the syntax of FA. You can find that here:
 - https://github.com/FAForever/lua-lang
 
 It is used at the moment to perform a basic syntax check for the Lua files of this repository.
 
 I've tried myself in the past to generate Visual Studio Code snippets. You can find that repository here:
  - https://github.com/FAForever/Snippet-Generator
  
It can successfully generate snippets that are useful for making campaign-related content. But it is not feasible for working in the repository. Some functions are available in one file, but not in another. The global context of sim related files is different from the global context of ui related files, as an example 😄 .

**Comment by hahn-kev** [comment]
Awesome, this sounds like it'll be a fun project.

The lua-lang repo will be helpful as a reference, but since it's C based it would probably be more work to use it than to just modify an existing extension already made for vscode. To start off I'll make a repo for it under my account and publish it as a vscode extension under my own account also, mostly just because that's the most convenient for me, however if you'd really like to under the FAForever github org I'm totally fine with that I'd just need access to the new repo. Let me know if it matters to you.

**Comment by Garanas** [comment]
> may I ask which extension do you currently use for Lua in VScode?

The ones described in the [instructions](https://github.com/FAForever/fa/blob/deploy/fafdevelop/setup/setup-english.md), which is essentially none that are directly related to Lua.

>The lua-lang repo will be helpful as a reference, but since it's C based it would probably be more work to use it than to just modify an existing extension already made for vscode. 

Sounds good to me. There's a history / changelog in the adjusted lua distribution that keeps track of what was changed. That should give you an initial pointer on what is different. 

As I have no experience with a functional Lua extension, could you give me an idea what we can expect? And what you would find interesting to work on? As an example, would the custom import functionality work? Would it provide some form of intellisense, or snippets?

> To start off I'll make a repo for it under my account and publish it as a vscode extension under my own account also, mostly just because that's the most convenient for me, however if you'd really like to under the FAForever github org I'm totally fine with that I'd just need access to the new repo. Let me know if it matters to you.

I've discussed this with another contributor. As long as the repository has an open source friendly license then that is fine to us. At some point I would prefer if it could become part of the FAForever group. That keeps all the software in one place. We can do that by forking your repository as time passes 😄

**Comment by hahn-kev** [comment]
Makes sense to me. Actually I've had some good luck with tweaking some settings and making a plugin for this [extension](https://marketplace.visualstudio.com/items?itemName=sumneko.lua) which is very feature rich. However it's a bit of a hack to get it working even this far so I may still fork it, I'm not sure yet. I'm still looking into it however. 

He's an example showing autocomplete from an import
![image](https://user-images.githubusercontent.com/4575355/168264326-21d34a95-3009-4422-8746-4d8c2cbffac5.png)
As you can see it knows what functions are defined and how they can be called, so you'll get warnings if you're not calling something correctly. To do this I had to sort of hack together a solution plugin for the extension that rewrites the file as the extension sees to, so there's some weird artifacts from doing that which I would prefer to fix.

You can also use go to definition and find references in many places which also means you can rename functions and such which will rename across multiple files properly.

It also has the ability to let you annotate types manually, so for example that's something that we would want to do on all the sim functions so that if you're calling them with a string instead of a number you would get a warning for example.

One question I have realized so far is that I don't understand when or how we can tell if a file is imported globally or must be imported in the specific file you want to use it. For example the `import` function is defined in `import.lua` however it's available globally, but `/lua/utilities.lua` must be imported if you want to use them. Is there some pattern I'm missing? Unless it's a simple as folder based we may have to annotate if a file is either a global file or one that needs to be imported, I'm not sure which would be best at this point but it would probably be based on which is most common.

**Comment by Garanas** [comment]
This is already exciting 😄 !

> Is there some pattern I'm missing? Unless it's a simple as folder based we may have to annotate if a file is either a global file or one that needs to be imported, I'm not sure which would be best at this point but it would probably be based on which is most common.

There are a few initialization files that determine the global scope of all files. 
 - simInit: https://github.com/FAForever/fa/blob/deploy/fafdevelop/lua/simInit.lua
 - simInit, hooked: https://github.com/FAForever/fa/blob/deploy/fafdevelop/schook/lua/simInit.lua
 - userInit: https://github.com/FAForever/fa/blob/deploy/fafdevelop/lua/userInit.lua
 - userInit, hooked: https://github.com/FAForever/fa/blob/deploy/fafdevelop/schook/lua/userInit.lua
 - globalInit: https://github.com/FAForever/fa/blob/deploy/fafdevelop/lua/globalInit.lua
 - globalInit, hooked: https://github.com/FAForever/fa/blob/deploy/fafdevelop/schook/lua/GlobalInit.lua
 
The fact that these are hooked (like mods would) is an artifact that we're still in the process of removing from the repository. The global initialization file is run by both the UI and sim Lua context. The user init file is specifically for UI code. The sim initialization file is for specifically for sim code. 

These are combined with these files:
 - Core global C functions:  https://github.com/FAForever/fa/blob/deploy/fafdevelop/engine/Core.lua
 - Sim global C functions:  https://github.com/FAForever/fa/blob/deploy/fafdevelop/engine/Sim.lua
 - UI global C functions:  https://github.com/FAForever/fa/blob/deploy/fafdevelop/engine/User.lua
 
Where again the Core is shared between the sim / ui Lua context. The context depends on where the file is stored. You can safely assume that any file starting at `/lua/ui` or at `/lua/maui` is part of the UI context. Any other file can be considered a sim file. This isn't entirely correct - but it is for the vast majority of cases.

> It also has the ability to let you annotate types manually, so for example that's something that we would want to do on all the sim functions so that if you're calling them with a string instead of a number you would get a warning for example.

That would be great - a lot of work but great. What type of format would that require?

**Comment by hahn-kev** [comment]
Alright, I can now say with confidence that I can get this to work for you. It does require building a binary to ship the extension, so I might take a few shortcuts on that and provide the extension as a file you can download and install manually, but it would only work for windows unless there's a big need for other platforms, but I'd need help testing that.

The only thing I really have to do before I can show it off is some minor tweaks to how it handles import paths. Right now it wants the imports to not start with a `/`` but we start them all with that as far as I can tell. Maybe I'm wrong though? What are some common import paths we will see? can I get some examples.

As for the annotations required. There's 2 ways I've set it up. Either you declare global files with `---@declare-global` or you declare imported files with `---@export-env`. There's a setting to toggle which you would like, but it makes more sense to me to declare the global files instead of the ones to export (which will then be imported somewhere else). This is a hard requirement, it was hard enough to get this to work, I don't think I can do it any other way.

Basically the annotations will configure the parser to treat something like this:
```lua
function myFunc() 
end
```
as either a function that's declared globally, or that it's returned from the file in a table for use in an import somewhere else (depending on the settings/annotations). The reason is that typical lua supports require modules which would be declared like this
```lua
function m.myFunc()
end
return {myFunc = myFunc}
```
the parser then picks that up and sees what's getting returned from the file when it's imported. I had to hack the parser to thing that the first example is like the previous one.

So I should be able to get that last hurdle covered and then send you the extension to test out, I'll contact you via discord.

**Comment by KionX** [comment]
local Control = import('control.lua').Control
local ScaleNumber = import('layouthelpers.lua').ScaleNumber

**Comment by Ejsstiil** [comment]
Maybe if we can set the root folder for project. In the .editorconfig perhaps... That way all paths with starting slash can be "absolute" inside VScode project
